internal/pkg/service/stream/storage/repository/volume/repo.go

```diff
@@ -0,0 +1,92 @@
+package volume
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	etcd "go.etcd.io/etcd/client/v3"
+
+	"github.com/keboola/keboola-as-code/internal/pkg/log"
+	"github.com/keboola/keboola-as-code/internal/pkg/service/common/etcdop"
+	"github.com/keboola/keboola-as-code/internal/pkg/service/common/etcdop/iterator"
+	"github.com/keboola/keboola-as-code/internal/pkg/service/common/etcdop/op"
+	"github.com/keboola/keboola-as-code/internal/pkg/service/common/etcdop/serde"
+	"github.com/keboola/keboola-as-code/internal/pkg/service/common/servicectx"
+	"github.com/keboola/keboola-as-code/internal/pkg/service/stream/storage/level/local/volume/assignment"
+	volume "github.com/keboola/keboola-as-code/internal/pkg/service/stream/storage/level/local/volume/model"
+	"github.com/keboola/keboola-as-code/internal/pkg/service/stream/storage/repository/volume/schema"
+)
+
+// Repository provides database operations with the storage.Metadata entity.
+// The orchestration of these database operations with other parts of the platform is handled by an upper facade.
+type Repository struct {
+	logger  log.Logger
+	process *servicectx.Process
+	client  *etcd.Client
+	schema  schema.Volume
+	volumes *etcdop.Mirror[volume.Metadata, volume.Metadata]
+}
+
+type dependencies interface {
+	Logger() log.Logger
+	Process() *servicectx.Process
+	EtcdClient() *etcd.Client
+	EtcdSerde() *serde.Serde
+}
+
+func NewRepository(d dependencies) (*Repository, error) {
+	r := &Repository{
+		logger:  d.Logger().WithComponent("volume.repository"),
+		process: d.Process(),
+		client:  d.EtcdClient(),
+		schema:  schema.New(d.EtcdSerde()),
+	}
+
+	if err := r.watchVolumes(); err != nil {
+		return nil, err
+	}
+
+	return r, nil
+}
+
+func (r *Repository) watchVolumes() error {
+	ctx, cancel := context.WithCancel(context.Background())
+	wg := &sync.WaitGroup{}
+	r.process.OnShutdown(func(ctx context.Context) {
+		cancel()
+		r.logger.Info(ctx, "closing volumes stream")
+		wg.Wait()
+		r.logger.Info(ctx, "closed volumes stream")
+	})
+
+	var errCh <-chan error
+	r.volumes, errCh = etcdop.SetupFullMirror(r.logger, r.schema.WriterVolumes().GetAllAndWatch(ctx, r.client)).StartMirroring(ctx, wg)
+
+	return <-errCh
+}
+
+// AssignVolumes assigns volumes to a new file.
+func (r *Repository) AssignVolumes(cfg assignment.Config, fileOpenedAt time.Time) assignment.Assignment {
+	return assignment.VolumesFor(r.volumes.All(), cfg, fileOpenedAt.UnixNano())
+}
```

Comment:
`Repository.AssignVolumes`  connets together well tested `assignment.VolumesFor`, and well tested `r.volumes` etcd mirror. 

E2E tests for the volumes assignment are part of the next file repository PR.